libs/utils/src/sync/gate.rs

```diff
@@ -102,48 +129,88 @@ impl Gate {
     /// important that the holders of such guards are respecting a CancellationToken which has
     /// been cancelled before entering this function.
     pub async fn close(&self) {
-        warn_if_stuck(self.do_close(), &self.name, Duration::from_millis(1000)).await
+        let started_at = std::time::Instant::now();
+        let mut do_close = std::pin::pin!(self.do_close());
+
+        let nag_after = Duration::from_secs(1);
+
+        let Err(_timeout) = tokio::time::timeout(nag_after, &mut do_close).await else {
+            return;
+        };
+
+        tracing::info!(
+            gate = ?self.as_ptr(),
+            elapsed_ms = started_at.elapsed().as_millis(),
+            "closing is taking longer than expected"
```

Comment:
This log line would benefit from keeping the original `name` field -- as written, if we have a GateGuard that blocks forever, we'll never know which object was affected, because we only see the span containing e.g. tenant/timeline ID when the GateGuard eventually drops.

Reply:
My thinking is that `close` will only ever be called within a spanned context which has tenant_id, timeline_id if not it's a bug because we are then probably losing a lot more of logging. See https://github.com/neondatabase/neon/pull/6542#issuecomment-1919009256 for example.

Reply:
My thinking:

1. find the slow thing (notice it from this info!)
2. then search for the pointer (unique within pageserver session)
3. find the culprit or restart pageserver

Reply:
That's reasonable -- might be worth checking the main shutdown paths to check that they have the right spans: e.g. looking at where Tenant::shutdown calls Timeline::shutdown, it reads to me like only the tenant Id is in the span.